### ISSUE TITLE
Storable won't read from a tie'd file handle

### ISSUE THREAD

**Issue by tlhackque** [Perl/perl5#19217]
Module: Storable

**Description**
If `Storable::fd_retrieve` is supplied with a non-`PerlIO` file handle (e.g. a `tie`'d handle from `IO::Uncompress::Unzip`), it fails to read the data, complaining about the magic number.

**Steps to Reproduce**
A detailed analysis and small reproducer can be found in [This IO::Uncompress::Unzip issue](https://github.com/pmqs/IO-Compress/issues/33)

**Expected behavior**
`Storable` should be able to read from (and write to) any valid Perl `FileHandle`, including those implelmented as `tie`'d objects.

The referenced issue asks that `IO::Uncompress::Unzip` provide a `PerlIO` handle (and has some ugly work-arounds).

However, there are many other `tie`'d objects that provide the `FILEHANDLE` interface; teaching `Storable` to accept it would fix all of those, as well as be consistent with the Perl DWIM standard...

As far as I can tell, `Storable` doesn't require an `PerlIO`-unique functions - the reproducer shows that supporting `OPEN` and `READ` (actually, `PerlIO`'s `FILL`) suffices for `fd_retrieve` - I haven't looked into `store_fd`, but assume it's equally undemanding.  Certainly, a fast path for `PerlIO` objects may be desirable.  But `tie`d objects should work.

**Perl configuration**
This fails with the latest `Storable` (3.25), Perl as old as 5.8.8, thru current.

**Comment by Gara312** [comment]
> Модуль: сохраняемый
> 
> **Описание** Если `Storable::fd_retrieve`предоставляется не `PerlIO`файловый дескриптор (например `tie`, дескриптор 'd from `IO::Uncompress::Unzip`), он не может прочитать данные, жалуясь на магическое число.
> 
> **Шаги по воспроизведению** Подробный анализ и небольшой репродуктор можно найти в [This IO :: Uncompress :: Unzip issue](https://github.com/pmqs/IO-Compress/issues/33)
> 
> **Ожидаемое поведение** `Storable` должно иметь возможность читать (и записывать) любой допустимый Perl `FileHandle`, включая те, которые реализованы как `tie`объекты 'd.
> 
> Указанная проблема требует `IO::Uncompress::Unzip`предоставления `PerlIO`дескриптора (и имеет некоторые уродливые обходные пути).
> 
> Однако есть много других `tie`объектов d, которые предоставляют `FILEHANDLE`интерфейс; обучение `Storable`его принятию исправит все это, а также будет соответствовать стандарту Perl DWIM ...
> 
> Насколько я могу судить, `Storable`не требует `PerlIO`Уникально функции - репродуктор показывает , что поддержка `OPEN`и `READ`( на самом деле, `PerlIO`«s `FILL`) достаточно для того `fd_retrieve`- я не смотрел на `store_fd`, но предположим , что это столь же нетребовательны. Конечно, `PerlIO`может быть желателен быстрый путь для объектов. Но `tie`объекты d должны работать.
> 
> **Конфигурация Perl.** Это не удается с последней `Storable`версией (3.25), Perl от версии 5.8.8 до текущей.

**Comment by tlhackque** [comment]
Chasing different issue, I came across another documented instance of `Storable` not accepting non-`PerlIO` filehandles [here](https://rt.cpan.org/Public/Bug/Display.html?id=23419). where the issue was first raised in 2006.

**Comment by Leont** [comment]
Yeah, it's using low-level IO that will never work with ties. Fixing this will require rewriting a bunch of this logic.
